### PR TITLE
refactor: introduce XFREE_CLEAR()

### PR DESCRIPTION
**Pull request by justinmk**
ref #1375

-  ~~rename `xfree` to `xfree_` (so that other PRs are forced to update upon merging to master)~~ 
-  ~~introduce `xfree`  macro (from @ZyX-I in #1375)~~
-  ~~rename `xfree` macro to `XFREE`~~

## Conclusion:

- see https://github.com/neovim/neovim/pull/10046#issuecomment-494572846
- introduce `XFREE_CLEAR()` and use it selectively

### CODE REVIEW

**Comment by bfredl** [comment]
If I understand correctly this is for robustness (null deref UB is less bad than use-after-free UB in practice) rather than semantics (as the code doesn't set it to NULL in ASAN build, code cannot rely on it)?

**Comment by justinmk** [comment]
@bfredl Yes. But I think the ASAN case might need to be removed, since it could result in different codepaths.

**Comment by bfredl** [comment]
@justinmk I also though of that. An alternative could be to set the pointer to 0x1 or something in a stable build, that would still preserve NULL comparison dependent code paths with the ASAN build. Preserving original use-after-free errors on ASAN build would be very useful..

**Comment by justinmk** [comment]
> @justinmk I also though of that. An alternative could be to set the pointer to 0x1 or something in a stable build, that would still preserve NULL comparison dependent code paths with the ASAN build. Preserving original use-after-free errors on ASAN build would be very useful..

Funny: with 0x1, some tests that were failing are now OK. I guess this means in many places the code compares a freed pointer to NULL. Not sure what to think about that...

**Comment by bfredl** [comment]
>  Funny: with 0x1, some tests that were failing are now OK. I guess this means in many places the code compares a freed pointer to NULL. Not sure what to think about that...

That is a related (potential) worry. If we do the change (as it was before, 0x0 not 0x1), future code _will_ start to relay on xfree:d pointer being nulled (regardless if we like it or not, no test will reveal it, so it will happen), which will make it harder to change back later. In that sense 0x1 is a more conservative change (if not entirely, non-null comparisons can still happen, but that should be much rarer).

**Comment by justinmk** [comment]
> non-null comparisons can still happen, but that should be much rarer

Seems to be quite widespread, based on the failed tests before the change to `0x01`. It's not clear to me whether either case is a problem in theory. But the current dependence on checking freed pointers for non-null is unfeasible to fix.

**Comment by bfredl** [comment]
> Seems to be quite widespread, based on the failed tests before the change to 0x01

@justinmk No, I meant comparisons that are not null-comparisons, i e comparisons between (now freed) object A and object B (which could be an alias of A), neither which is NULL. Those will be revealed as failures that remain even in the 0x1 build.

File: src/nvim/macros.h [hunk 1]
```diff
@@ -198,6 +198,16 @@
 # define IO_COUNT(x)  (x)
 #endif
 
+#define xfree(ptr) \
```
Comment (jamessan) [inline src/nvim/macros.h]: Shouldn't this be `XFREE` so it's a) clear it is a macro and b) _actually_ forces people to handle the merge properly (as you hinted at initially).

**Comment by justinmk** [comment]
Going to retreat from doing this indiscriminately, because comparing pointers after freeing them is a common pattern. Example in `tv_list_remove_items()`:

```c
    xfree(li);
    if (li == item2) {
      break;
    }
```

Instead we can do it selectively/explicitly via `XFREE_CLEAR()`.

File: src/nvim/macros.h (hunk 1 above)
Comment (justinmk) [reply src/nvim/macros.h]: yeah, it was WIP. But I'm retreating anyway :)

**Comment by janlazo** [comment]
What about 8.0.1481?

File: src/nvim/lib/kvec.h [hunk 2]
```diff
@@ -58,7 +58,7 @@
     }
 
 #define kv_init(v) ((v).size = (v).capacity = 0, (v).items = 0)
-#define kv_destroy(v) xfree((v).items)
+#define kv_destroy(v) XFREE_CLEAR((v).items)
```
Comment (bfredl) [inline src/nvim/lib/kvec.h]: it seems weird to only reset the pointer and not the other fields: the data structure is now in an inconsistent state. Maybe `(xfree((v).items), kv_init(v))`.

Comment (justinmk) [reply src/nvim/lib/kvec.h]: c4f56c4

**Comment by janlazo** [comment]
Is 8.0.1481 N/A now?

**Comment by justinmk** [comment]
It was covered by https://github.com/neovim/neovim/pull/10046/commits/ae846b41dfed16446be6469cb01f12f1eb1fa534 . Did you find something missing?

**Comment by janlazo** [comment]
I didn't see that https://github.com/neovim/neovim/commit/ae846b41dfed16446be6469cb01f12f1eb1fa534 has `vim-patch:8.0.1481` until now. I compared the changes now and found nothing missing.

**Pull request merged**